File: LeetCode/Array/Golang/Leetcode655-print-binary-tree.go
package sbgo

import "strconv"

func printTree(root *TreeNode) [][]string {
	hei := 0
	var dfs func(node *TreeNode, level int)
	dfs = func(node *TreeNode, level int) {
		if node == nil {
			return
		}
		if level > hei {
			hei = level
		}
		dfs(node.Left, level+1)
		dfs(node.Right, level+1)
	}
	dfs(root, 0)
	n, m := hei+1, (1<<(hei+1))-1
	ans := make([][]string, n)
	for i := 0; i < n; i++ {
		ans[i] = make([]string, m)
	}
	var dfs_fill func(node *TreeNode, x, y int)
	dfs_fill = func(node *TreeNode, x, y int) {
		if node == nil {
			return
		}
		ans[x][y] = strconv.Itoa(node.Val)
		if node.Left != nil {
			dfs_fill(node.Left, x+1, y-(1<<(hei-x-1)))
		}
		if node.Right != nil {
			dfs_fill(node.Right, x+1, y+(1<<(hei-x-1)))
		}
	}
	dfs_fill(root, 0, (m-1)/2)
	return ans
}

File: LeetCode/Array/JS/Leetcode655-print-binary-tree.js
var printTree = function(root) {
    let hei = 0;
    let dfs = function(node,level) {
        if(!node) {
            return;
        }
        if(level > hei) {
            hei = level;
        }
        dfs(node.left,level + 1);
        dfs(node.right,level + 1); 
    }
    dfs(root,0);
    const [n,m] = [hei + 1,(1 << (hei + 1)) - 1]
    let ans = new Array(n).fill(0).map(()=> new Array(m).fill(""));
    let dfs_fill = function(node,x,y) {
        if(!node) {
            return;
        }
        ans[x][y] = String(node.val);
        if(node.left) {
            dfs_fill(node.left,x + 1,y - (1 << (hei - x - 1)));
        }
        if(node.right) {
            dfs_fill(node.right,x + 1,y + (1 << (hei - x - 1)));
        }
    }
    dfs_fill(root,0,parseInt((m - 1) / 2));
    return ans;
};